src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFilter.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement.FeatureFilters
+{
+    /// <summary>
+    /// A feature filter that can be used to activate features for targeted audiences.
+    /// </summary>
+    [FilterAlias(Alias)]
+    public class ContextualTargetingFilter : IContextualFeatureFilter<ITargetingContext>
+    {
+        private const string Alias = "Microsoft.Targeting";
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="loggerFactory">A logger factory for creating loggers.</param>
+        public ContextualTargetingFilter(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory?.CreateLogger<ContextualTargetingFilter>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Performs a targeting evaluation using the provided <see cref="TargetingContext"/> to determine if a feature should be enabled.
+        /// </summary>
+        /// <param name="context">The feature evaluation context.</param>
+        /// <param name="targetingContext">The targeting context to use during targeting evaluation.</param>
+        /// <returns>True if the feature is enabled, false otherwise.</returns>
+        public Task<bool> EvaluateAsync(FeatureFilterEvaluationContext context, ITargetingContext targetingContext)
+        {
+            TargetingFilterSettings settings = context.Parameters.Get<TargetingFilterSettings>() ?? new TargetingFilterSettings();
+
+            //
+            // No audience targeted, the feature will be off
+            if (settings.Audience == null)
+            {
+                _logger.LogWarning($"The '{Alias} feature filter does not have an audience for the feature '{context.FeatureName}'.");
+
+                return Task.FromResult(false);
+            }
+
+            //
+            // Check if the user is being targeted directly
+            if (targetingContext.UserId != null &&
+                settings.Audience.Users != null &&
+                settings.Audience.Users.Any(user => targetingContext.UserId.Equals(user)))
+            {
+                return Task.FromResult(true);
+            }
+
+            //
+            // Check if the user is in a group that is being targeted
+            if (targetingContext.Groups != null &&
+                settings.Audience.Groups != null)
+            {
+                foreach (string group in targetingContext.Groups)
```

Comment:
>foreach [](start = 16, length = 7)

This might confuse people by the order we search. For example in audience:
Group A -> 80%
Group B -> 20 %
and in user context we have Groups -> [ B, A ]
We will match 20% here but user might expect 80% since audience order is A first. 
Which I can expect if you have group A and you should match group A first instead of group B in audience order.

Reply:
The order doesn't matter though. There would be no distinguishable difference. We test both groups and if either one should result in the feature being on then it will be on.